synapse/storage/databases/main/user_directory.py

```diff
@@ -241,8 +242,11 @@ def _get_next_batch(
                     or await self.should_include_local_user_in_dir(user_id)
                 }
 
-                # Update each user in the user directory.
+                # Upsert a user_directory record for each remote user we see.
+                # (Local users are processed in `_populate_user_directory_users`.)
                 for user_id, profile in users_with_profile.items():
+                    if self.hs.is_mine_id(user_id):
+                        continue
```

Comment:
it might be worth adding a comment here to explain why it's important to skip local users here.